CONTRIBUTING.md

```diff
@@ -0,0 +1,109 @@
+   adds your name to the [`CONTRIBUTORS`](CONTRIBUTORS) file.
+3. Follow our [Coding Guidelines](#coding-guidelines).
+4. Expect that a reviewer will ask you for restructuring your commits! This
+   usually happens towards the end of the lifetime of a PR when it matured
+   enough for merging.
+
+# Coding Guidelines
+
+These guidelines are horribly incomplete at this point in time, but one has to
+start somewhere ;) Exactly because these guidelines are relatively new, do not be
+alarmed if large parts of this repo do not follow them (yet). The idea is to
+apply them incrementally on new code and when existing code is modified. These
+guidelines will also evolve as time passes.
+
+How to reference a rule in an issue or PR discussion:
+```
+// reference to our guideline
+[G.2](https://github.com/FairRootGroup/FairRoot/blob/master/CONTRIBUTING.md#g.2-adopted-c++-core-guidelines)
+// reference to c++ core guideline with prefix CPPCG::
+[CPPCG::I.11](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#i11-never-transfer-ownership-by-a-raw-pointer-t-or-reference-t)
+```
+
+## G - General
+
+### G.1 Our CI runs [`clang-tidy`](https://clang.llvm.org/extra/clang-tidy/)
+
+**Config**: [`.clang-tidy`](.clang-tidy)
+
+We currently do not strictly enforce *clang-tidy* checks, but the author
+should try to address any warnings. The set of enabled checks will evolve over
+time.
+
+### G.2 Adopted C++ Core Guidelines
+
+This shall be an evolving list of explicitely adopted C++ Core Guidelines:
+
+* [CPPCG::I.11: Never transfer ownership by a raw pointer (`T*`) or reference (`T&`)](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#i11-never-transfer-ownership-by-a-raw-pointer-t-or-reference-t)
+   * If an owning raw pointer cannot be avoided for legacy reasons,
+     **you must add a comment documenting the ownership semantics!**
+
+### G.3 Write a good Git history
+   * Follow [the seven rules of a great Git commit message](https://cbea.ms/git-commit/#seven-rules)!
+   * Use a meaningful commit granularity, e.g. do not mix
+     unrelated changes in a single commit and vice verse squash related commits
+     into one.
+   * Utilize [the commit message body to explain *what* and *why* and not *how*](https://cbea.ms/git-commit/#why-not-how).
+     Aim your message to be meaningful (concise, but complete
+     enough) to a reader in the future!
+   * Utilize [reference keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
+     both in commit messages as well as in PR comments if applicable.
+
+## D - Documentation
+
+### D.1 Use [doxygen-style comments](https://www.doxygen.nl/manual/commands.html)
+
+## S - Style
+
+### S.1 Our CI enforces formatting with [`clang-format`](https://clang.llvm.org/docs/ClangFormat.html)
+
+**Config**: [`.clang-format`](.clang-format)
```

Comment:
You should exactly define which version is used in the CI and probably also fix  a version in the CI.
Unfortunately the changes produced by clang-format sometimes changes slightly from version to version.
We had this problem with CbmRoot in the beginning and where only able to solve it by providing a static version of clang-format for download which can be used practically on any Linux system. We also have a static binary for download for macosx .

Reply:
We had this exactly twice till now. And it boiled down to non precise declarations in `.clang-format`. Being more explicit in there fixed it for those cases.

See: ab58c0136e9f2a671365ebbb6a329b9c477c25ed, 3d7b969d6fbed0dabcf76ce17f8c9107bba3d8f1

That said, I guess, we haven't officially settled on something with this. Maybe the answer currently is: The clang-format that comes with the latest ubuntu (any, not only LTS) release. Or more generally: The CI is the authoritative version.

Reply:
It is fine to define clang-format to be the latests version. I checked for Ubuntu 22.10 the version is 15. How do I use it as Debian 10 or 11 user? On this systems the version is 7 respectively 11.

Reply:
I think all the discussions boil down to the question how we deal with failing clang-format tests. If we require that they must be fixed by the user we need to have a way that they can do this easily independent on their system.

Reply:
I see a few points here:

* Usually the current `.clang-format` produces the exact same result on Debian-11's clang-format (I am using that on my Desktop). And any differences–as already noted–usually boil down to imprecise declarations in `.clang-format` and should be fixed.

  That said, the differences are usually very minor and can easily be fixed by looking at the CI output and applying the changes manually. (If they're big, you could even cut'n'paste them into `patch`.)

* If you actually think, that this should be *very* easy for users, our CI could open a "reverse" Pull Request on the submitters branch and "suggest" the needed changes. But really, the changes are quite minor currently.

Reply:
As a side note: In the FairMQ we opted to **not require** the format check. But this only works, if reviewers stay on top of this check and *care* about formatting.

> if you actually think, that this should be very easy for users, our CI could open a "reverse" Pull Request on the submitters branch and "suggest" the needed changes. But really, the changes are quite minor currently.

While it is fancy, I disliked such a feature so far (whenever I encountered it in other projects) because it likely leads to worse git history with tons of bot-authored, auto-generated "fix format" commits.

Reply:
I totally agree with trying to avoid bot authored commits in the final history. Reviewers should request a squash in that case.

And really, I currently really doubt that the investment into such a bot would pay off (as already explained).